### PR TITLE
feat: add fixed size binary encoding to lance

### PR DESCRIPTION
**Pull request by raunaks13**
Addresses #2717
- Supports nullable, binary, string, large string, and large binary types that are fixed-size. Whether the type is fixed-size or not is detected automatically.
- Arrays with empty strings, as well as empty arrays use the old binary encoder
- New protobuf contains the byte width (int) and bytes (encoded using value encoder). Offsets are created in memory during decode time (should save an IOP since we don't encode the offsets).

Initial benchmarking results:
Full scan, fixed size string of length 8, 1M rows
Screenshot 2024-08-13 at 3 58 20 PM
To reproduce: `pytest python/benchmarks/test_fixed_size_binary.py`.

### CODE REVIEW

**Comment by broccoliSpicy** [comment]
fixed length binary is quite common in traditional row oriented database systems such as mysql/postgres `char`, I don't know about the arrow community, do you think there is any chance that they will accept a proposal of such a datatype?

in mysql and postgres, the length of fixed length is a upper bound and paddings are added,
`If the string to be stored is shorter than the declared length, values of type character will be space-padded;`[postgres](https://www.postgresql.org/docs/current/datatype-character.html)
`The length of a CHAR column is fixed to the length that you declare when you create the table. The length can be any value from 0 to 255. When CHAR values are stored, they are right-padded with spaces to the specified length. When CHAR values are retrieved, trailing spaces are removed unless the [PAD_CHAR_TO_FULL_LENGTH](https://dev.mysql.com/doc/refman/8.4/en/sql-mode.html#sqlmode_pad_char_to_full_length) SQL mode is enabled.`[mysql](https://dev.mysql.com/doc/refman/8.4/en/char.html), 
if arrow won't add a fixed-width data type, do you think we can find a algorithm to determine the upper bound and enable fixed-length binary encoding(with padding) when it is a good case to apply such a encoding?

**Comment by raunaks13** [comment]
> I don't know about the arrow community, do you think there is any chance that they will accept a proposal of such a datatype?

I believe arrow does have a [fixed size binary datatype](https://arrow.apache.org/docs/cpp/api/datatype.html#classarrow_1_1_fixed_size_binary_type). We can use this directly.
Currently the idea is that we'll use the fixed size binary encoding for fixed size binary datatypes, as well as binary data that happens to be fixed size, and is perhaps smaller than a certain size threshold.

**Comment by broccoliSpicy** [comment]
ha, arrow does have a fixed size binary datatype.

> Currently the idea is that we'll use the fixed size binary encoding for fixed size binary datatypes, as well as binary data that happens to be fixed size, and is perhaps smaller than a certain size threshold.

this sounds like a very good approach, 👍

File: rust/lance-encoding/src/encoder.rs [hunk 1]
```diff
@@ -430,6 +475,83 @@ fn check_dict_encoding(arrays: &[ArrayRef], threshold: u64) -> bool {
     true
 }
 
+fn check_fixed_size_encoding(arrays: &[ArrayRef], byte_width: &mut u64) -> bool {
+    // check if all arrays in arrays have the same length
```
Comment (westonpace) [inline rust/lance-encoding/src/encoder.rs:479]:
```suggestion
    // check if all items in arrays have the same length
```

File: rust/lance-encoding/src/encoder.rs [hunk 2]
```diff
@@ -291,6 +293,8 @@ impl CoreArrayEncodingStrategy {
         data_type: &DataType,
         data_size: u64,
         use_dict_encoding: bool,
+        use_fixed_size_encoding: bool,
+        byte_width: &u64,
```
Comment (westonpace) [inline rust/lance-encoding/src/encoder.rs:297]: This feels like too many arguments to this method.  I'll comment an alternative later on.

File: rust/lance-encoding/src/encoder.rs [hunk 3]
```diff
@@ -430,6 +475,83 @@ fn check_dict_encoding(arrays: &[ArrayRef], threshold: u64) -> bool {
     true
 }
 
+fn check_fixed_size_encoding(arrays: &[ArrayRef], byte_width: &mut u64) -> bool {
```
Comment (westonpace) [inline rust/lance-encoding/src/encoder.rs:478]: Output parameters should generally be avoided if possible.  They are sometimes inevitable (e.g. in a visitor pattern / recursive calling pattern) but I don't think this one is.  You could return `Option<u64>`.

File: rust/lance-encoding/src/encoder.rs [hunk 4]
```diff
@@ -430,6 +475,83 @@ fn check_dict_encoding(arrays: &[ArrayRef], threshold: u64) -> bool {
...
+        } else if let Some(arr) = arr.as_string_opt::<i64>() {
+            arr.iter().flatten().all(|s| !s.is_empty())
+        } else if let Some(arr) = arr.as_binary_opt::<i64>() {
+            arr.iter().flatten().all(|s| !s.is_empty())
+        } else {
+            panic!("wrong dtype");
+        }
+    }) {
+        return false;
+    }
```
Comment (westonpace) [inline rust/lance-encoding/src/encoder.rs:499]: Why is this needed if you have a check later on to filter out lengths of zero anyways?  Ideally we only iterate though all values in the array a single time.

File: rust/lance-encoding/src/encoder.rs [hunk 5]
```diff
@@ -443,10 +565,24 @@ impl ArrayEncodingStrategy for CoreArrayEncodingStrategy {
         let data_type = arrays[0].data_type();
         let use_dict_encoding = data_type == &DataType::Utf8
             && check_dict_encoding(arrays, get_dict_encoding_threshold());
+
+        let binary_datatypes = [
+            DataType::Utf8,
+            DataType::LargeUtf8,
+            DataType::Binary,
+            DataType::LargeBinary,
+        ];
```
Comment (westonpace) [inline rust/lance-encoding/src/encoder.rs:574]: Should be a const.

File: rust/lance-encoding/src/encoder.rs [hunk 6]
```diff
@@ -443,10 +565,24 @@ impl ArrayEncodingStrategy for CoreArrayEncodingStrategy {
...
+
+        let byte_width = &mut 0;
+        let use_fixed_size_encoding =
+            binary_datatypes.contains(data_type) && check_fixed_size_encoding(arrays, byte_width);
```
Comment (westonpace) [inline rust/lance-encoding/src/encoder.rs:578]: It's ok to use the arrays to check if an encoding is applicable.  We need to do that for both dictionary and here for fixed size string.  These are unlike the other encodings where we just use the type.

The way it's coded up though requires passing these `use_xyz_encoding` args to `array_encoder_from_type` which is a little awkward.  I could envision two approaches.

First approach.  If we've determined that these encodings are correct, why even bother calling `array_encoder_from_type`?  We should be able to just new up the encoding and return it.

Second approach, if some encodings need the arrays (and not just the data type) then we could change `array_encoder_from_type` to `choose_array_encoder` and pass in the arrays.  Something like...

```
// Gather common info needed by all encodings
let data_size = arrays
    .iter()
    .map(|arr| arr.get_buffer_memory_size() as u64)
    .sum::<u64>();
let data_type = arrays[0].data_type();
// Pick an encoding
Self::choose_array_encoder(data_type, data_size, arrays, self.version, Some(&field.metadata))
```

In other words, we rename `array_encoder_from_type` as `choose_array_encoder`.  Then the "does dictionary apply" or "does fixed size apply" can go into the same method as all the other "does xyz encoder apply" logic.

File: rust/lance-encoding/src/encodings/physical/fixed_size_binary.rs [hunk 7]
```diff
@@ -0,0 +1,388 @@
...
+    async fn test_fixed_size_large_binary() {
+        let field = Field::new("", DataType::LargeBinary, true);
+        check_round_trip_encoding_random(field, HashMap::new()).await;
+    }
+
+    #[test_log::test(tokio::test)]
+    async fn test_fixed_size_large_utf8() {
+        let field = Field::new("", DataType::LargeUtf8, true);
+        check_round_trip_encoding_random(field, HashMap::new()).await;
+    }
```
Comment (westonpace) [inline rust/lance-encoding/src/encodings/physical/fixed_size_binary.rs:256]: How are you generating fixed size arrays here?  Aren't these tests just using the normal binary encoding?

File: rust/lance-encoding/src/encodings/physical/fixed_size_binary.rs [hunk 8]
```diff
@@ -0,0 +1,388 @@
...
+        let big_string = String::from_iter((0..(1000 * 1000)).map(|_| '0'));
+        let string_array = StringArray::from_iter_values((0..90).map(|_| big_string.clone()));
+
+        check_round_trip_encoding_of_data(
+            vec![Arc::new(string_array)],
+            &TestCases::default(),
+            HashMap::new(),
+        )
+        .await;
+    }
```
Comment (westonpace) [inline rust/lance-encoding/src/encodings/physical/fixed_size_binary.rs:352]: This doesn't seem to be testing the fixed size encoding (and we probably don't need to worry about the jumbo case for this encoding anyways)

File: rust/lance-encoding/src/encodings/physical/fixed_size_binary.rs [hunk 9]
```diff
@@ -0,0 +1,388 @@
...
+        // When encoding an array of empty strings there are no bytes to encode
+        // which is strange and we want to ensure we handle it
+        let string_array = Arc::new(StringArray::from(vec![Some(""), None, Some("")]));
+
+        let test_cases = TestCases::default().with_range(0..2).with_indices(vec![1]);
+        check_round_trip_encoding_of_data(vec![string_array.clone()], &test_cases, HashMap::new())
+            .await;
+        let test_cases = test_cases.with_batch_size(1);
+        check_round_trip_encoding_of_data(vec![string_array], &test_cases, HashMap::new()).await;
+    }
```
Comment (westonpace) [inline rust/lance-encoding/src/encodings/physical/fixed_size_binary.rs:367]: We don't actually use fixed size encoding here and we have empty string encodings elsewhere.  I don't think this test is needed.

File: rust/lance-encoding/src/encodings/physical/fixed_size_binary.rs [hunk 10]
```diff
@@ -0,0 +1,388 @@
...
+        for _ in 0..5000 {
+            string_builder.append_option(Some(&giant_string));
+        }
+        let giant_array = Arc::new(string_builder.finish()) as ArrayRef;
+        let arrs = vec![giant_array];
+
+        // // We can't validate because our validation relies on concatenating all input arrays
+        let test_cases = TestCases::default().without_validation();
+        check_round_trip_encoding_of_data(arrs, &test_cases, HashMap::new()).await;
+    }
```
Comment (westonpace) [inline rust/lance-encoding/src/encodings/physical/fixed_size_binary.rs:387]: Same as above, no need to worry about jumbo strings for FSB.

File: protos/encodings.proto [hunk 11]
```diff
@@ -218,6 +218,11 @@ message PackedStruct {
   Buffer buffer = 2;
 }
 
+message FixedSizeBinary {
+  ArrayEncoding bytes = 1;
+  uint32 byte_width = 2;
+}
```
Comment (westonpace) [inline protos/encodings.proto:224]: Why is a new message needed?  How is this different from `Flat`?

**Review by westonpace — COMMENTED** [review]
Some questions from a first pass but looks cool.  Thanks for creating this.

File: rust/lance-encoding/src/encodings/physical/fixed_size_binary.rs (hunk 7 above)
Comment (raunaks13) [reply rust/lance-encoding/src/encodings/physical/fixed_size_binary.rs:256]: I think this test actually generates fixed size arrays internally...check `rand_utf8()` in `generator.rs`
I also printed which encoding is used to double check. We should probably update the test, actually (maybe in a separate PR)

File: rust/lance-encoding/src/encoder.rs (hunk 4 above)
Comment (raunaks13) [reply rust/lance-encoding/src/encoder.rs:499]: We need to distinguish between null values and empty strings (both end up having offset lengths = 0). Since we want the fixed size encoding to support nulls, but we want to use the binary encoding for empty strings. So this check is to filter out the empty ones, and the one later handles the nulls.

File: rust/lance-encoding/src/encodings/physical/fixed_size_binary.rs (hunk 10 above)
Comment (raunaks13) [reply rust/lance-encoding/src/encodings/physical/fixed_size_binary.rs:387]: Sure. Although again, I think this will end up using the fixed size encoding because it is essentially encoding a bunch of strings of the same size. So if we keep the test in `binary.rs`, we should probably modify it.

File: protos/encodings.proto (hunk 11 above)
Comment (raunaks13) [reply protos/encodings.proto:224]: Flat uses a buffer, but we need an ArrayEncoding? Since bytes is actually a BasicEncoder(ValueEncoder())
Otherwise it's hard to reconstruct the BasicPageScheduler directly? We would need to reconstruct the ArrayEncoding from the buffer...

File: rust/lance-encoding/src/encoder.rs (hunk 6 above)
Comment (raunaks13) [reply rust/lance-encoding/src/encoder.rs:578]: I don't think we can use this approach directly
For example, the `use_dict_encoding` flag needs to be computed for arrays with a binary datatype.
And now for the inner bytes we need to figure out which encoding to use (fixed size binary, binary). This requires calling `choose_array_encoder` again. In this case, if we don't pass this parameter recursively, we may end up using the dict encoding again and get a stack overflow.

**Comment by westonpace** [comment]
> if arrow won't add a fixed-width data type, do you think we can find a algorithm to determine the upper bound and enable fixed-length binary encoding(with padding) when it is a good case to apply such a encoding?

@broccoliSpicy I do want to get to this eventually.  For now, we are just using it when a page has strings that are **exactly** the same length (no padding needed).  Also, no arrow type is strictly needed since this is for storage only.  So this is just an opportunistic encoding like dictionary that we detect and apply if possible.

**Review by westonpace — CHANGES_REQUESTED** [review]
Let's gate this behind 2.1.  This will both preserve our current testing (since the other tests will be using 2.0 and not use this encoding) and help keep 2.0 stable as we're about to make it the default.

File: rust/lance-encoding/src/encoder.rs (hunk 6 above)
Comment (raunaks13) [reply rust/lance-encoding/src/encoder.rs:578]: We don't compute the actual dictionary in the beginning, so `arrays` will still be the parent array and can have duplicates as per the current structure
For now, I have restructured the `use_fixed_size_encoding()` check and moved it inside `choose_array_encoder`. 
I agree this "choose encoder" workflow probably needs more work in a follow-up.

**Review by westonpace — APPROVED** [review]

**Comment by broccoliSpicy** [comment]
Not sure if we can also do this, 
in postgresql's `char`, when storage layer pass data to compute layer, the trailing spaces the user input has is also striped away.
Screenshot 2024-08-19 at 11 09 13 AM
